File: hackerrank/coin_change.py
def getWays(n, c):
    def helper(target, coins, memo={}):

        if target == 0:
            return 1
        if target < 0:
            return 0

        if (target, len(coins)) in memo:
            return memo[(target, len(coins))]

        total = 0
        for i in range(len(coins)):
            coin = coins[i]
            if coin > target:
                continue
            else:
                total += helper(target - coins[i], coins[i:], memo)

        memo[(target, len(coins))] = total

        return memo[(target, len(coins))]

    return helper(n, c)


print(getWays(4, [1, 2, 3]))

File: hackerrank/coin_change.js
var change = function(amount, coins) {
    function helper(a, c, memo = {}){
        if (a == 0){
            return 1
        };
        if(a < 0){
            return 0
        };
        var name = a + "#" +c.length;
        if (name in memo){

            return memo[name]
        };

        var total = 0;
        for(var i = 0; i < c.length; i++){

            var coin = c[i];
            if (coin > a){
                continue;
            };
             total += helper(a - coin, c.slice(i), memo);
        };
        memo[name] = total
        return memo[name]
    };
    return helper(amount, coins);
};